### DISCUSSION TITLE
pyinstalerr not add .dat file

### DISCUSSION THREAD

**Discussion by Coolhatcker** [pyinstaller/pyinstaller#6416] [Help]
Hello. I used command pyinstaller --add-data "TurboActivate.dll;" --add-data "TurboActivate.dat ;." example.py for create exe file, but after convertation exe file cant find .dat file, only .dll. Please help

**Comment by rokm** [comment]
Are both files collected in `dist\example` and your python code fails to find the `.dat` file, or is the `.dat` file not collected at all?

**Reply by Coolhatcker** [reply]
I dont know. I think the .dat file not collected at all, just .dll file... and i dont understand why
But i can see in spec datas=[('TurboActivate.dll', '.'),  ('TurboActivate.dat', '.')],

**Reply by rokm** [reply]
Can you show us the generated .spec file?

**Reply by Coolhatcker** [reply]
```
`# -*- mode: python ; coding: utf-8 -*-


block_cipher = None


a = Analysis(['index_exe.py'],
             pathex=[],
             binaries=[],
             datas=[('new-map.log', '.'), ('TurboActivate.dll', '.'), ('TurboActivate.dat', '.')],
             hiddenimports=[],
             hookspath=[],
             hooksconfig={},
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)

exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,  
          [],
          name='index_exe',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          upx_exclude=[],
          runtime_tmpdir=None,
          console=True,
          disable_windowed_traceback=False,
          target_arch=None,
          codesign_identity=None,
          entitlements_file=None , icon='BC.ico')
`
```

**Reply by rokm** [reply]
> I think the .dat file not collected at all

Can you check? I.e., by looking in the `dist\example` directory?

**Reply by Coolhatcker** [reply]
I use --onefile

**Reply by rokm** [reply]
Then check if you see it listed in `build\example\PKG-00.toc`.

**Reply by Coolhatcker** [reply]
Hmm. The file is present in the PKG-00.toc

**Reply by rokm** [reply]
Then it should be collected. How are you trying to access it?

**Reply by Coolhatcker** [reply]
I used this instruction https://wyday.com/limelm/help/using-turboactivate-with-python/

**Reply by rokm** [reply]
The instructions are for `onedir` build, though, not `onefile` one...

If you look at `turboactivate` code ([here](https://github.com/wyday/python-turboactivate/blob/b905958338790fec0802bc2586801c7a00a7669d/turboactivate/__init__.py#L51-L79)), you'll see that unless you explicitly provide it with location of the data file, it expects to find it next to the executable. Which is not true for `onefile` builds, because the bundle files are extracted in temporary location.

So you should pass the path to the data file (i.e., `os.path.join(sys._MEIPASS, 'TurboActivate.dat')`) via `dat_file_loc` argument when you instantiate your `TurboActivate` instance.

**Reply by Coolhatcker** [reply]
Thank you so much!! Can try

**Reply by Coolhatcker** [reply]
Used `dat_file_loc = os.path.join(sys._MEIPASS, 'TurboActivate.dat')` when instantiate TurboActivate instance, but its didn't work...

**Reply by rokm** [reply]
In that case, you'll have to figure out yourself what went wrong. Perhaps first try to get things working with a `onedir` build, and once that works, move back to `onefile`. Maybe add a `print` statement to `turboactivate` code before it tries to open the dat file to print the filepath it is trying to open; and maybe add an `input()` to your code to pause execution, so you can go inspect the temporary directory and check if the .dat file is where it is supposed to be...
